plugin/clone/include/clone_common.h

```diff
@@ -25,7 +26,37 @@ class Ha_clone_common_cbk : public Ha_clone_cbk {
  public:
   [[nodiscard]] int precopy(THD *thd, uint task_id) override;
 
-  [[nodiscard]] int synchronize_engines() override;
+ protected:
+  /** Perform the cross-engine synchronization for logs: execute a
+    performance_schema.log_status query, and call set_log_stop for each storage
+    engine with its part of STORAGE_ENGINES column JSON object from that query.
+    Also assign synchronization coordinates get from log_status table to the
+    parameter synchronization_coordinates.
+    @param[in]	synchronization_coordinates	syncronization coordinates
+    @return error code */
+  int synchronize_logs(Key_Values &synchronization_coordinates);
+
+ private:
+  /** Get json object from the json wrapper after doing json validations
+    @param[in]	object_name the object name, used for debug printing
+    @param[in]	json_wrapper json wrapper to be converted to be json object
+    @return converted json object */
+  const Json_object *get_json_object(const char *object_name,
+                                     Json_wrapper &json_wrapper);
+
+  /** Extract key and values from local_repl_info.
+      local_repl_info has 3 pieces of information: gtid, binlog_file, offset.
+      We will also get gtid corresponding to the binlog_file and offset and
+    record that gtid as well. This is to sanity check that this gtid we get from
+    binlog_file and offset is same as what we have in local_repl_info.
+    @param[in]	local_repl_info the "local" column we get from log_status table
+    @param[in]	synchronization_coordinates Key_Values of gtid(from log_status
+    table), binlog_file, binlog_offset and gtid(from binlog_file and
+    binlog_offset)
+    @return error code */
+  int populate_synchronization_coordinates(
+      const Json_object *local_repl_info,
```

Comment:
Here and everywhere else: replace pointer args with reference args whenever possible (i.e. the pointer cannot be nullptr, no arithmetics done with it). This applies to out parameters too